src/site/content/en/reliable/adaptive-loading-sw/index.md

```diff
@@ -0,0 +1,166 @@
+
+```javascript
+import {registerRoute} from 'workbox-routing';
+import {CacheFirst} from 'workbox-strategies';
+
+registerRoute(
+  new RegExp('/img/'),
+  new CacheFirst()
+);
+```
+
+Caching strategies can be customized with [Workbox plugins](https://developers.google.com/web/tools/workbox/guides/using-plugins). These allow you to add additional behaviors by manipulating requests and responses during the lifecycle of a request. Workbox has several built-in plugins for common cases and APIs, but you can also define a [custom plugin](https://developers.google.com/web/tools/workbox/guides/using-plugins#custom_plugins), and introduce some custom logic of your choice.
+
+To achieve adapting loading, define a custom plugin, called, for example, `adaptiveLoadingPlugin`:
+
+```javascript
+const adaptiveLoadingPlugin = {
+  requestWillFetch: async ({request}) => {
+    const urlParts = request.url.split('/');
+    let imageQuality;
+
+    switch (
+      navigator && navigator.connection
+        ? navigator.connection.effectiveType
+        : ''
+    ) {
+      //...
+      case '3g':
+        imageQuality = 'q_30';
+        break;
+      //...
+    }
+
+    const newUrl = urlParts
+      .splice(urlParts.length - 1, 0, imageQuality)
+      .join('/')
+      .replace('.jpg', '.png');
+    const newRequest = new Request(newUrl.href, {headers: request.headers});
+
+    return newRequest;
+  },
+};
+```
+
+The previous code does the following:
+
+- Implements a `requestWillFetch()` callback: This is called whenever a network request is about to be made, so you can alter the [Request](https://developer.mozilla.org/en-US/docs/Web/API/Request).
+- Checks the connection type, by using the [Network Information API](https://developer.mozilla.org/en-US/docs/Web/API/NavigatorConcurrentHardware/hardwareConcurrency). Based on the status of the network, it creates a new URL part, indicating the quality of the image to fetch (e.g. `q_30` for 3G users).
+- Creates a new URL based on the dynamic newPart value, and returns the new [Request](https://developer.mozilla.org/en-US/docs/Web/API/Request) to be made, based on that URL.
+
+Next, pass the plugin to a cacheFirst strategy containing a regular expression to match image URLs (e.g. `/img/`):
+
+```javascript/5
+workbox.routing.registerRoute(
+  new RegExp('/img/'),
+  workbox.strategies.cacheFirst({
+    cacheName: 'images',
+    plugins: [
+      adaptiveLoadingPlugin,
+      workbox.expiration.Plugin({
+        maxEntries: 50,
+        purgeOnQuotaError: true,
+      }),
+    ],
+  }),
+);
+```
+
+As a result, when requests for images are intercepted, the runtime caching strategy will try to fulfill the request from the cache. If it's not available, it will run the logic in the plugin, to decide which image quality to fetch from the network.
+
+Finally the response will be persisted in the cache, and sent back to the page.
+
+## Cloudinary Workbox Plugin
```

Comment:
@demianrenzulli & @jeffposnick I'm just realizing that this section is going against our policy of singling out commercial services. We'll leave it for now but we may have problems in the future and might need to revisit it.

Reply:
Thanks for bringing this up. We were not aware of this policy in Web.dev., but it's good to keep it in mind for future posts as well. If you think we have to make any changes, please, let us know!